### PR TITLE
#83100 Use mocked HttpSender for unit tests

### PR DESCRIPTION
**Pull request by tabman83**

### CODE REVIEW

File: src/Tests/CaptainHook.Tests/Web/WebHooks/GenericWebhookHandlerTests.cs [hunk 1]
```diff
@@ -122,13 +122,14 @@ public async Task ExecuteHappyPathRawContract_DeliveryFailure()
             var mockHttp = new MockHttpMessageHandler();
             var webhookRequest = mockHttp.When(HttpMethod.Put, $"{config.Uri}/{metaData["OrderCode"]}")
                 .WithContentType("application/json", messageData.Payload)
-                .Respond(HttpStatusCode.ServiceUnavailable, "application/json", string.Empty);
+                .Respond(HttpStatusCode.OK, "application/json", string.Empty);
```
Comment (chomick) [inline src/Tests/CaptainHook.Tests/Web/WebHooks/GenericWebhookHandlerTests.cs:125]: Did we not abstract Http client now? Should it not be httpSender that is mocked now?
That will significant;y simplify those tests.

On top of that, you can have unit tests for http-sender where http client mock would be used.

Comment (tabman83) [reply src/Tests/CaptainHook.Tests/Web/WebHooks/GenericWebhookHandlerTests.cs:125]: That may have slipped through, as I aimed at using the mocked httpsender for all of them

File: src/Tests/CaptainHook.Tests/Web/WebHooks/GenericWebhookHandlerTests.cs [hunk 2]
```diff
@@ -93,7 +93,7 @@ public async Task ExecuteHappyPathRawContract()
 
         [IsUnit]
         [Fact]
-        public async Task ExecuteHappyPathRawContract_DeliveryFailure()
+        public async Task WebhookConfig_MapsCorrectlyToHttpCall()
```
Comment (chomick) [inline src/Tests/CaptainHook.Tests/Web/WebHooks/GenericWebhookHandlerTests.cs:96]: The way those unit tests were developed originally is a mess.
Why are they in the "Web" folder?
What exactly are they testing (what is the unit here)?
IMP, they have to be refactored and files/folders renamed to follow some logic/structure.

At minimum, the test projects should map 1-2-1 with the actual projects. They should follow the same folder structure, with an option to have them broken down even further into separate files, if necessary.

Can you create a tech debt story for that (along with suggestions what & how has to be refactored), please, and when this PR/story is done, can we move to the new one?

File: src/CaptainHook.EventHandlerActor/Handlers/HttpSender.cs [hunk 3]
```diff
@@ -1,4 +1,4 @@
-﻿using CaptainHook.Common;
+using CaptainHook.Common;
```
Comment (chomick) [inline src/CaptainHook.EventHandlerActor/Handlers/HttpSender.cs:1]: What's the reason this was included?

File: src/Tests/CaptainHook.Tests/Web/WebHooks/WebhookResponseHandlerTests.cs [hunk 4]
```diff
@@ -285,22 +295,31 @@ public async Task BadCheckMultiRouteSelection()
             var messageData = CreateMessageDataPayload();
             messageData.SubscriberConfig = EventHandlerConfigWithBadMultiRoute;
 
-            var mockHttpHandler = new MockHttpMessageHandler();
-            mockHttpHandler.When(HttpMethod.Post, expectedWebHookUri)
-                .WithContentType("application/json; charset=utf-8", expectedContent)
-                .Respond(HttpStatusCode.OK, "application/json", "{\"msg\":\"Hello World\"}");
+            //var mockHttpHandler = new MockHttpMessageHandler();
```
Comment (chomick) [inline src/Tests/CaptainHook.Tests/Web/WebHooks/WebhookResponseHandlerTests.cs:298]: Any reason this code has to stay in the file?

File: src/Tests/CaptainHook.Tests/Web/WebHooks/WebhookResponseHandlerTests.cs [hunk 5]
```diff
@@ -242,36 +253,35 @@ public async Task GoodCheckMultiRouteSelection()
...
-            }
+            mockHttpSender
+                .Setup(x => x.SendAsync(messageData.SubscriberConfig.Callback.HttpMethod, new Uri(ExpectedCallbackUri), It.IsAny<WebHookHeaders>(),
+                    It.IsAny<string>(), It.IsAny<TimeSpan>(), It.IsAny<CancellationToken>()))
+                .ReturnsAsync(new HttpResponseMessage(HttpStatusCode.OK)
+                {
+                    Content = new StringContent(ResponseContentMsgHelloWorld, System.Text.Encoding.UTF8, "application/json")
+                });
 
             var mockBigBrother = new Mock<IBigBrother>();
```
Comment (chomick) [inline src/Tests/CaptainHook.Tests/Web/WebHooks/WebhookResponseHandlerTests.cs:276]: We should have those mocks created in the ctor

File: src/Tests/CaptainHook.Tests/Web/WebHooks/WebhookResponseHandlerTests.cs [hunk 6]
```diff
@@ -206,29 +207,39 @@ public async Task CheckHookFailureDoesNotInvokeCallback()
             var messageData = CreateMessageDataPayload();
             messageData.SubscriberConfig = SubscriberConfigurationWithSingleRoute;
 
-            var mockHttpHandler = new MockHttpMessageHandler();
-            mockHttpHandler.When(HttpMethod.Post, expectedWebHookUri)
-                .WithContentType("application/json; charset=utf-8", expectedContent)
-                .Respond(HttpStatusCode.ServiceUnavailable, "application/json", "{\"msg\":\"Hello World\"}");
+            //var mockHttpHandler = new MockHttpMessageHandler();
```
Comment (chomick) [inline src/Tests/CaptainHook.Tests/Web/WebHooks/WebhookResponseHandlerTests.cs:210]: Any reason we leave this code here?

File: src/Tests/CaptainHook.Tests/Web/WebHooks/WebhookResponseHandlerTests.cs [hunk 7]
```diff
@@ -242,36 +253,35 @@ public async Task GoodCheckMultiRouteSelection()
...
             var mockAuthHandlerFactory = CreateMockAuthHandlerFactory();
-            var webhookResponseHandler = BuildWebhookResponseHandler(messageData, httpClients, mockBigBrother, mockAuthHandlerFactory);
+            var webhookResponseHandler = BuildWebhookResponseHandler(messageData, mockHttpSender, mockBigBrother, mockAuthHandlerFactory);
 
             await webhookResponseHandler.CallAsync(messageData, new Dictionary<string, object>(), _cancellationToken);
 
             mockAuthHandlerFactory.Verify(e => e.GetAsync(It.IsAny<WebhookConfig>(), _cancellationToken), Times.AtMostOnce);
-
-            Assert.Equal(1, mockHttpHandler.GetMatchCount(multiRouteCall));
+            mockHttpSender.Verify(x => x.SendAsync(HttpMethod.Post, new Uri(expectedMultiWebhookUri), It.IsAny<WebHookHeaders>(), It.IsAny<string>(), It.IsAny<TimeSpan>(), It.IsAny<CancellationToken>()), Times.Once);
```
Comment (chomick) [inline src/Tests/CaptainHook.Tests/Web/WebHooks/WebhookResponseHandlerTests.cs:283]: For verification, should you not check that correct parameters are passed, incl. headers, timeout etc?

**Review by chomick — APPROVED** [review]
Significant refactoring of the unit tests in this area is required. Nothing blocking this PR in my opinion but this is far from perfect.
As I mentioned below, we need to create a tech-debt to refactor tests in this area (the whole 'web' folder has to be redone properly and distributed across multiple projects/folders)

File: src/CaptainHook.EventHandlerActor/Handlers/HttpSender.cs (hunk 3 above)
Comment (nikhilgirraj) [reply src/CaptainHook.EventHandlerActor/Handlers/HttpSender.cs:1]: For the `MessageData` class, it is used as payload for testing `CallAsync` method in webhook response handler

File: src/Tests/CaptainHook.Tests/Web/WebHooks/WebhookResponseHandlerTests.cs (hunk 4 above)
Comment (nikhilgirraj) [reply src/Tests/CaptainHook.Tests/Web/WebHooks/WebhookResponseHandlerTests.cs:298]: Nope. Removed it.

File: src/Tests/CaptainHook.Tests/Web/WebHooks/WebhookResponseHandlerTests.cs (hunk 5 above)
Comment (nikhilgirraj) [reply src/Tests/CaptainHook.Tests/Web/WebHooks/WebhookResponseHandlerTests.cs:276]: Moved.

File: src/Tests/CaptainHook.Tests/Web/WebHooks/WebhookResponseHandlerTests.cs (hunk 6 above)
Comment (nikhilgirraj) [reply src/Tests/CaptainHook.Tests/Web/WebHooks/WebhookResponseHandlerTests.cs:210]: Mistake, removed.

File: src/Tests/CaptainHook.Tests/Web/WebHooks/WebhookResponseHandlerTests.cs (hunk 7 above)
Comment (nikhilgirraj) [reply src/Tests/CaptainHook.Tests/Web/WebHooks/WebhookResponseHandlerTests.cs:283]: Not sure what those values should be, I can check the code and specify that.

**Review by dawidkeshop — APPROVED** [review]

File: src/CaptainHook.EventHandlerActor/Handlers/HttpSender.cs (hunk 3 above)
Comment (chomick) [reply src/CaptainHook.EventHandlerActor/Handlers/HttpSender.cs:1]: But there is no other change in this file - only using? did it fail to build before?

**Pull request merged**